### PR TITLE
Earthquake V2 contracts overhaul

### PR DESCRIPTION
**Pull request by 3xHarry**

### CODE REVIEW

File: src/v2/Carousel/Carousel.sol [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+    {
+        emissionsToken = IERC20(_emissionsToken);
+        relayerFee = _relayerFee;
+        feeTreasury = _feeTreasury;
+    }
+
+    /*///////////////////////////////////////////////////////////////
+                        DEPOSIT/WITHDRAWAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
```
Comment (neverDefined) [inline src/v2/Carousel/Carousel.sol:68]: natspec for consistency

File: src/v2/Carousel/Carousel.sol [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+        epochIdExists(_id)
+        epochHasNotStarted(_id)
+        nonReentrant
+    {
+        if (_receiver == address(0)) revert AddressZero();
+
+        _asset().safeTransferFrom(msg.sender, address(this), _assets);
+
+        // mint logic, either in queue or direct deposit
+        if (queueClosed(_id)) {
```
Comment (neverDefined) [inline src/v2/Carousel/Carousel.sol:86]: maybe you want to have lower and higher bound for the queue closed, like when it is `late` and `impossible`

File: src/v2/Carousel/Carousel.sol [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+                index++;
+            }
+
+            rolloverAccounting[_epochId] = index;
+
+            asset.safeTransfer(feeTreasury, _operations * relayerFee);
+        }
+    }
+
+    function queueClosed(uint256 _epochId) public view returns (bool) {
```
Comment (neverDefined) [inline src/v2/Carousel/Carousel.sol:362]: might be cleaner: 

if (epoch 0)...
if (block.timestamp)..
return false

File: src/v2/Carousel/Carousel.sol [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+            QueueItem[] memory queue = rolloverQueue;
+            uint256 length = rolloverQueue.length;
+            // revert if queue is empty or operations are more than queue length
+            if (length == 0 || _operations > length - 1) revert OverflowQueue();
+            // account for how many operations have been done
+            uint256 index = rolloverAccounting[_epochId];
+            while (index < _operations) {
+                // only roll over if user won last epoch
+                if (
+                    previewWithdraw(queue[index].epochId, queue[index].assets) >
```
Comment (neverDefined) [inline src/v2/Carousel/Carousel.sol:334]: might be cleaner to abstract this check to, `hasUserWonEpoch(id)` (would be useful external read too

File: src/v2/Carousel/Carousel.sol [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+    }
+
+    modifier notRollingOver(
+        address _receiver,
+        uint256 _epochId,
+        uint256 _assets
+    ) {
+        if (ownerToRollOverQueueIndex[_receiver] != 0) {
+            QueueItem memory item = rolloverQueue[getRolloverIndex(_receiver)];
+            if (item.epochId == _epochId && item.assets < _assets)
```
Comment (3xHarry) [inline src/v2/Carousel/Carousel.sol:473]: should only be able to withdraw  balance[epoch] - items.asset

**Pull request closed without merging**